mrp_unbuild_rebuild_variant/models/mrp_unbuild_rebuild_variant.py

```diff
@@ -0,0 +1,223 @@
+
+    name = fields.Char(
+        default=lambda self: self.env["ir.sequence"].next_by_code(self._name),
+        readonly=True,
+    )
+    state = fields.Selection(
+        [("draft", "Draft"), ("done", "Done")],
+        required=True,
+        default="draft",
+        readonly=True,
+    )
+    # TODO: WHAT SHOULD BE COMPUTED HERE
+    unbuild_product_id = fields.Many2one("product.product")
+    unbuild_bom_id = fields.Many2one("mrp.bom", compute="_compute_unbuild_product_id")
+    unbuild_location_id = fields.Many2one("stock.location")
+    # if unbuild product is tracked:
+    unbuild_lot_id = fields.Many2one(
+        "stock.production.lot", domain="[('product_id', '=', unbuild_product_id)]"
+    )
+    # if unbuild product's bom includes tracked components:
+    unbuild_original_mo_id = fields.Many2one("mrp.production")
+    # if unbuild product's bom does not include tracked components:
+    quantity = fields.Float()
+
+    rebuild_product_template = fields.Many2one(
+        "product.template", compute="_compute_valid_template"
+    )
+
+    # TODO: Domain for same product.template
+    rebuild_product_id = fields.Many2one(
+        "product.product",
+        # FIXME: domain="[
+        # FIXME:    ('product_tmpl_id', '=', 'unbuild_product_id.product_tmpl_id.id')
+        # FIXME: ]"
+    )
+    # TODO: Check if this bom could be different than unbuild_bom_id
+    rebuild_bom_id = fields.Many2one("mrp.bom", compute="_compute_rebuild_product_id")
+    rebuild_location_id = fields.Many2one("stock.location")
+    rebuild_lot_id = fields.Many2one("stock.production.lot")
+
+    unbuild_order_id = fields.Many2one("mrp.unbuild", readonly=True)
+    rebuild_order_id = fields.Many2one("mrp.production", readonly=True)
+
+    @api.onchange("rebuild_product_id")
+    def _compute_rebuild_product_id(self):
+        for record in self:
+            if record.rebuild_product_id:
+                record.rebuild_bom_id = record.env["mrp.bom"]._bom_find(
+                    product=record.rebuild_product_id
+                )
+
+    @api.onchange("unbuild_product_id")
+    def _compute_unbuild_product_id(self):
+        for record in self:
+            if record.unbuild_product_id:
+                record.unbuild_bom_id = record.env["mrp.bom"]._bom_find(
+                    product=record.unbuild_product_id
+                )
+
+    def _filter_bom_lines_for_variant(self, lines, variant):
+        """
+        Filter lines with no attributes or
+        where product and line attributes intersection is not null.
+        """
+        attributes = variant.attribute_value_ids
+        lines = lines.filtered(
+            lambda l: not l.attribute_value_ids or attributes & l.attribute_value_ids
+        )
+        return lines
+
+    def _check_availability(self):
+        """
+        Checks that all products to create the `rebuild product` are available
+        either in the `unbuild product` (by dismantling it) or in the stock.
```

Comment:
For this first version, we should ensure the components needed to rebuild the product are generated by the unbuild order only.
The issue in taking something from the stock, is that we don't have any handle to select a lot number if the missing component is tracked. So we cannot force an assignation without the possibility for the user to select which lot should be taken from the stock.

This could be mentioned in the ROADMAP, as it can be improved later on.

Reply:
What I check here is that there's enough available products (once the `unbuild product` is dismantled).
Later on, I just ensure that each common component between `unbuild product` and `rebuild product` share the same lot number.
For any other component, this is the default odoo behaviour.